File: src/parse.js
/**
 * Extracts a JSON object from raw model output.
 * Handles: markdown code fences, leading/trailing prose,
 * and the model occasionally being chatty despite instructions.
 * Throws a descriptive error if no parseable JSON is found.
 */
export function extractJson(rawText) {
    // --- Temp Debug ---
    console.log("=== extractJson received ===");
    console.log("\nRAW TEXT");
    console.log(rawText);
    console.log("END RAW TEXT\n");
    console.log("typeof:", typeof rawText);
    console.log("length:", rawText?.length);
    console.log("JSON.stringify of input:", JSON.stringify(rawText));
    console.log("=== end ===");
    // --- End Temp Debug ---

    if (typeof rawText !== "string" || rawText.trim() === "") {
        throw new Error("extractJson: received empty or non-string input");
    }

    let text = rawText.trim();
    console.log("[1] after trim, length:", text.length, "has {:", text.includes("{"));

    // Only strip fences if the string ACTUALLY starts with a code fence.
    if (text.startsWith("```")) {
    // Remove opening fence line (``` or ```json) and closing fence
    text = text
        .replace(/^```(?:json)?\s*\n?/, "")
        .replace(/\n?```\s*$/, "")
        .trim();
    }

    // Locate the outermost JSON object by first { and last }
    const firstBrace = text.indexOf("{");
    const lastBrace = text.lastIndexOf("}");
    console.log("[4] firstBrace:", firstBrace, "lastBrace:", lastBrace);

    if (firstBrace === -1 || lastBrace === -1 || lastBrace < firstBrace) {
        throw new Error(
            `extractJson: no JSON object found in model output. Raw: ${rawText.slice(0, 200)}`
        );
    }

    const candidate = text.slice(firstBrace, lastBrace + 1);

    try {
        return JSON.parse(candidate);
    } catch (err) {
        throw new Error(
            `extractJson: found JSON-like text but failed to parse: ${err.message}. Candidate: ${candidate.slice(0, 200)}`
        );
    }
}

File: py/parse.py
import json
import re


def extract_json(raw_text: str) -> object:
    """
    Extracts a JSON object from raw model output.
    Returns object (unvalidated) — the caller is responsible for validation.
    Handles: markdown fences, leading/trailing prose, chatty models.

    Logic is identical to parse.ts — this is one of the places where
    the two languages look most similar.
    """
    if not isinstance(raw_text, str) or not raw_text.strip():
        raise ValueError("extract_json: received empty or non-string input")

    text = raw_text.strip()

    # Only strip fences if the string actually starts with one
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*\n?", "", text)
        text = re.sub(r"\n?```\s*$",          "", text)
        text = text.strip()

    first_brace = text.find("{")
    last_brace  = text.rfind("}")

    if first_brace == -1 or last_brace == -1 or last_brace < first_brace:
        raise ValueError(
            f"extract_json: no JSON object found in model output. Raw: {raw_text[:200]}"
        )

    candidate = text[first_brace : last_brace + 1]

    try:
        return json.loads(candidate)
    except json.JSONDecodeError as err:
        raise ValueError(
            f"extract_json: found JSON-like text but failed to parse: {err}. "
            f"Candidate: {candidate[:200]}"
        ) from err
